### PR TITLE
ci-operator: document operator-sdk testing

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @petr-muller @AlexNPavel 

https://server-ci-tools-1104.apps.ci.l2s4.p1.openshiftapps.com/help/operators

### CODE REVIEW

File: pkg/webreg/webreg.go [hunk 1]
```diff
@@ -401,6 +401,112 @@ const templateDefinitions = `
...
+    tag: "8"
+  operand:           # imports the latest operand image
+    namespace: "ocp"
+    name: "operand"
+    tag: "latest"
+images:
+- from: ubi
+  to: operator                         # the pipeline:operator image, built on top of ubi:8
+  dockerfile_path: Dockerfile
+- from: ubi
```
Comment (stevekuznetsov) [inline pkg/webreg/webreg.go:484]: imho we should not allow configuring the base image for a bundle?

File: pkg/webreg/webreg.go [hunk 2]
```diff
@@ -401,6 +401,112 @@ const templateDefinitions = `
...
+  to: operator-bundle                  # the pipeline:bundle image, build on top of ubi:8
+  dockerfile_path: bundle.Dockerfile
```
Comment (stevekuznetsov) [inline pkg/webreg/webreg.go:486]: imho we should default this if unset to the operator sdk default (`bundle.Dockerfile`)

File: pkg/webreg/webreg.go [hunk 3]
```diff
@@ -401,6 +401,112 @@ const templateDefinitions = `
...
+  operator_manifests: bundle/manifests # where operator manifests reside in the repository
```
Comment (stevekuznetsov) [inline pkg/webreg/webreg.go:487]: imho we should default this if unset to the operator sdk default (`bundle/manifests`)

File: pkg/webreg/webreg.go (hunk 1 above)
Comment (petr-muller) [reply pkg/webreg/webreg.go]: that's probably correct

File: pkg/webreg/webreg.go [hunk 4]
```diff
@@ -401,6 +401,112 @@ const templateDefinitions = `
...
+  operand:           # imports the latest operand image
+    namespace: "ocp"
+    name: "operand"
+    tag: "latest"
+images:
+- from: ubi
+  to: operator                         # the pipeline:operator image, built on top of ubi:8
+  dockerfile_path: Dockerfile
+- from: ubi
+  to: operator-bundle                  # the pipeline:bundle image, build on top of ubi:8
```
[the remark this answers is not in the document]
Comment (petr-muller) [reply pkg/webreg/webreg.go]: Unless you have a repo with multiple operators, each of which will need a bundle. I was told that this is possible.

File: pkg/webreg/webreg.go (hunk 2 above)
Comment (petr-muller) [reply pkg/webreg/webreg.go]: yeah, why not

File: pkg/webreg/webreg.go (hunk 3 above)
Comment (petr-muller) [reply pkg/webreg/webreg.go]: this looks like a pitfall to me though, I don't think I've ever seen a real operator repo that would have a simple path like this

File: pkg/webreg/webreg.go [hunk 5]
```diff
@@ -401,6 +401,112 @@ const templateDefinitions = `
...
+repository, an index image will automatically be built to package those bundles.
+</p>
+
+<h3 id="tests"><a href="#tests">Running Tests</a></h3>
+
+<p>
+Once <code>ci-operator</code> builds the operator bundle and index, they are
+available to be used as a <code>CatalogSource</code> by OLM for deploying and
+testing the operator. Today, the only style of testing that's supported is an
+end-to-end test, implemented with the <a href="/workflow/optional-operators-cvp-common-aws"><code>optional-operators-cvp-common-aws</code></a>
```
[the remark this answers is not in the document]
Comment (petr-muller) [reply pkg/webreg/webreg.go:448]: No, and the end to end workflow is still basically a nop, AFAIK. Plus, currently the built artifacts will not be exposed in the variables that the subscribe step consumes - that wiring is still missing.

**Comment by stevekuznetsov** [comment]
@AlexNPavel @petr-muller updated, redeployed the doc -- we need to hash out the workflow to use but this should otherwise be ready

File: pkg/webreg/webreg.go [hunk 6]
```diff
@@ -401,6 +401,116 @@ const templateDefinitions = `
...
+    tag: "latest"
+operator:
+  bundles: # entries create bundle images from Dockerfiles and an index containing all bundles
+  - dockerfile_path: path/to/Dockerfile # defaults to bundle.Dockerfile
+    context_dir: path/                  # defaults to .
+  substitutions:
+  - pullspec: "quay.io/openshift/operand:1.3"  # this will replace references to the operand with the imported version
+    with: "stable:operand"
+  - pullspec: "quay.io/openshift/operator:1.3" # this will replace references to the operator with the built version
+    with: "pipeline:ci-bundle-0"
```
Comment (AlexNPavel) [inline pkg/webreg/webreg.go:496]: `ci-bundle-0` is the bundle image, not the operator image. This should instead point to the operator image which would be in the `images` section in the config.

File: pkg/webreg/webreg.go [hunk 7]
```diff
@@ -401,6 +401,116 @@ const templateDefinitions = `
...
+`
+
+const optionalOperatorTestConfig = `tests:
+- as: "operator-e2e"
+  steps:
+    workflow: "optional-operators-cvp-common-aws"
```
Comment (AlexNPavel) [inline pkg/webreg/webreg.go:502]: There are currently 3 env vars in the workflow that are required according to the bash for the `optional-operators-subscribe` step: https://ci-operator-configresolver-ui-ci.apps.ci.l2s4.p1.openshiftapps.com/reference/optional-operators-subscribe#line7. I think this was created before env var support in step registry, so we'll need to update the `optional-operators-subscribe` step to specifically require them.

We should probably add the required env vars to the example (`OO_INDEX`, `OO_PACKAGE`, and `OO_CHANNEL`; we can probably update the workflow to not require the index though and just set that to `ci-index`).

**Pull request closed without merging**